### PR TITLE
[AMD] aiter: honor per-layer softmax scale

### PR DESCRIPTION
**Pull request by vpremseth23**
📄 **Design & context (all 5 PRs):** https://gist.github.com/vpremseth23/18ee4d3f8504dd553660bbe00e397e73 — combined rationale, architecture, before/after accuracy + speed/throughput benchmarks, and reproduction steps.

> **Stacked series — aiter/Gemma-4 on ROCm (1/5).** Base of the stack; independent, mergeable on its own.

## Motivation

The aiter attention backend hardcoded the softmax scale to `head_dim ** -0.5`
(or the precomputed `self.scale`) at five kernel call sites, ignoring
`layer.scaling`. That is correct only for models whose scale equals the default
and is **silently wrong** for any model that sets a custom scale — e.g. Gemma-4
uses `scaling=1`, so at `head_dim=256` the affected paths scaled attention logits
by `1/16` with no error or warning. `triton_backend` and `trtllm_mha_backend`
already honor `layer.scaling`; this brings aiter in line.

## Modifications

Pass `layer.scaling` instead of the hardcoded value at the five sites: prefill
(`mha_batch_prefill_func`), decode and paged decode (`paged_attention_ragged`),
and the two SHUFFLE-5D paths in `aiter_utils.py`. No-op for every model whose
scale is `head_dim ** -0.5` (the value `layer.scaling` already holds).

## Accuracy Tests

Hardware: AMD MI350X (gfx950), `--attention-backend aiter`, bf16 KV, TP=1.

**Capability — custom-scale model (`google/gemma-4-31b-it`, `scaling=1`):** the
correct scale is required for coherent output; with the full stack applied,

| Metric | Result | Threshold |
|---|---|---|
| GSM8K (n=200) | **0.825** (invalid 0.005) | ≥ 0.80, coherent |

Reverting this PR reintroduces the `1/16` logit scaling → incoherent output, so
the 0.825 result directly validates the fix.

**Non-regression — default-scale model (`Qwen/Qwen2.5-7B-Instruct`):**

| GSM8K (n=200) | before | after |
|---|---|---|
| aiter | **0.860** | **0.860** |

Identical (this PR is a no-op when `layer.scaling == head_dim ** -0.5`).

```bash
python -m sglang.test.few_shot_gsm8k --num-questions 200 --host 127.0.0.1 --port 30000
```

## Speed Tests and Profiling

Scalar argument change → no measurable perf effect. `Qwen/Qwen2.5-7B-Instruct`,
aiter, `random` 1024in/512out:

| Output tok/s | before | after |
|---|---|---|
| @concurrency 8 | 1689.21 | 1683.89 |
| @concurrency 128 | 11963.88 | 11918.02 |

Within run-to-run noise.

```bash
python -m sglang.bench_serving --backend sglang-oai --dataset-name random \
  --random-input 1024 --random-output 512 --random-range-ratio 1.0 \
  --num-prompts 1000 --max-concurrency 8 --host 127.0.0.1 --port 30000   # + 128
```

## Checklist

- [x] Pre-commit / code style clean.
- [x] Unit tests: behavior is a scalar passed into GPU kernels; validated by the
      accuracy tests above (Gemma-4 requires the correct scale; Qwen unchanged).
- [x] Documentation: n/a (no user-facing surface change).
- [x] Accuracy + speed benchmarks provided above.
- [x] Follows SGLang code style.

## Reproduction (common setup)

```bash
docker run -it --device=/dev/kfd --device=/dev/dri --security-opt seccomp=unconfined \
  --ipc=host --network host lmsysorg/sglang:v0.5.9-rocm700-mi35x
pip install -e /sgl-workspace/oss-sglang/python --no-deps   # this branch
python -m sglang.launch_server --model-path <model> --attention-backend aiter \
  --kv-cache-dtype auto --page-size 32 --tp 1 --trust-remote-code \
  --host 127.0.0.1 --port 30000
```

---
### CI States

Latest PR Test (Base): :x: [Run #34426857054](https://github.com/sgl-project/sglang/actions/runs/34426857054)
Latest PR Test (Extra): :x: [Run #34426856870](https://github.com/sgl-project/sglang/actions/runs/34426856870)
Latest PR Test (AMD ROCm 10): :x: [Run #34426857015](https://github.com/sgl-project/sglang/actions/runs/34426857015)

---
**Stacked series — aiter/Gemma-4 on ROCm (review/merge in order):**
1. #38754 — per-layer softmax scale
2. #38755 — cross-layer KV guard
3. #38756 — SWA KV-pool resolver + paged-decode guard
4. #38757 — unified_attention prefill (**depends on #38756**)
5. #38758 — allow aiter for Gemma-4 (**merge last; depends on #38754–#38757**)

### CODE REVIEW

**Comment by vpremseth23** [comment]
cc @kkHuang-amd @hubertlu-tw @1am9trash — this is an AMD/`aiter` (ROCm) change; your review would be appreciated. Design rationale, before/after accuracy + speed/throughput benchmarks (MI350X), and reproduction steps are linked at the top of the description. (@HaiShaw is already requested via CODEOWNERS.)

**Review by HaiShaw — APPROVED** [review]
